### DISCUSSION TITLE
how set NavigationMenu title

### DISCUSSION THREAD

**Discussion by wow-prime** [serenity-is/Serenity#7181] [Help] [answered]
like below picture:
![image](https://github.com/serenity-is/Serenity/assets/19904096/090a4960-93cf-4c3d-86e7-d2f5380f0f02)

I want set Title to NavigationMenu(set A to B).
How can I do it?

My code is in NavigationItems.cs.

public List\<NavigationItemAttribute\> GetItems()
{
    string strSystemCode = _config.GetValue<string>("SystemCode");
    var items = new List<NavigationItemAttribute> { };
    using (var connection = _sqlConnections.NewByKey("ePortal"))
    {
        var menu = connection.Query<SystemMenu>(@"select * from Menu where SystemId = (select SystemId from [System] where Code = @Code) and Code != '00' ",
            param: new
            {
                Code = strSystemCode
            },
            commandType: CommandType.Text).ToList();

        for(int i = 0; i < menu.Count(); i++)
        {
            string strPath = menu[i].CodeName;
            var itemSelf = menu[i];
            while (true)
            {
                var itemParent = menu.Find(x => x.Code == itemSelf.ParentCode);
                if (itemParent == null)
                    break;
                strPath = itemParent.CodeName + "/" + strPath;
                itemSelf = itemParent;
            }
            strPath = strSystemCode + "/" + strPath;
            int intOrder;

            if (menu.Find(x => x.ParentCode == menu[i].Code) == null)
                items.Add(new NavigationLinkAttribute(int.TryParse(menu[i].Order, out intOrder) ? intOrder : 999999,
                    path: strPath,
                    url: menu[i].Url,
                    permission: menu[i].CodeName + "_Read",// "*",
                    icon: "icon-speedometer"
                    ));
            else
                items.Add(new NavigationMenuAttribute(int.TryParse(menu[i].Order, out intOrder) ? intOrder : 999999,
                    title: strPath,
                    
                    //permission: menu[i].CodeName + "_Read",// "*",
                    icon: "icon-speedometer"
                    )); 
            items[i].Title = menu[i].Name;
        }
    }
    return items;
}

**Comment by LSNicholls** [comment]
To be honest, I'm having some difficulty understanding your code. 

At the beginning of the `for` loop through your menu items, you define the value of string `strPath = menu[i].CodeName`, then you potentially re-set it in an inner `while` loop (if no break happens), and then after the inner `while` loop you re-set it for *every* item (`strPath = strSystemCode + "/" + strPath`)?  Am I misreading this?

And, after that, you similarly seem to set the title attribute for Navigation Menu items in the `else` part of an `if/else` construct, but for the Navigation Link attributes, you don't set it, followed by setting ` items[i].Title = menu[i].Name` for *both* branches?

If I had to guess I would guess that the difference between your section headers (Navigation Menu) and your menu items (Navigation Link) results from the difference between the `if/else` blocks.   Maybe the ` items[i].Title = menu[i].Name` is setting the tooltip but the lower-case title attribute is setting the visual caption.

So, I'll ask a couple of things:

1. can you verify what the items[i].Title is doing, if anything?
2. what is the structure of your Menu table?
3. can you explain your thinking on the setting and re-setting of the strPath variable?

Also I want to say that I have never put code into NavigationItems.cs the way you are doing since the file appears to be declarative in nature.  Maybe you got this idea from the Dynamic Navigation sample?  In any case, to change the menu behavior, I have put code in NavigationModelFactory.cs, and there you are manipulating Navigation Items, not Menu or Link Item Attributes, regardless of which type it is.  The actual work of gathering both types of items is done by Navigation Helper.  Have you looked at https://github.com/serenity-is/Serenity/blob/master/src/Serenity.Net.Web/Navigation/NavigationHelper.cs ?

**Reply by wow-prime** [reply]
1. can you verify what the items[i].Title is doing, if anything?
  =>I set menuname to items[i].Title in NavigationLinkAttribute then menu show menuname 
  But I set menuname to items[i].Title in NavigationMenuAttribute then menu not show menuname. 

2. what is the structure of your Menu table?
 
![image](https://github.com/serenity-is/Serenity/assets/19904096/988b7f50-7a69-4fe9-9285-14ca9d7234e9)

3. can you explain your thinking on the setting and re-setting of the strPath variable?
 =>I want these menu in strSystemCode NavigationGroup

4. And, after that, you similarly seem to set the title attribute for Navigation Menu items in the else part of an if/else construct, but for the Navigation Link attributes, you don't set it, followed by setting  items[i].Title = menu[i].Name for both branches?
=>Navigation Link attributes no title attribute

5. Also I want to say that I have never put code into NavigationItems.cs the way you are doing since the file appears to be declarative in nature. Maybe you got this idea from the Dynamic Navigation sample? 
=> Yes, I got this idea from the Dynamic Navigation sample.

6.  I have put code in NavigationModelFactory.cs, and there you are manipulating Navigation Items, not Menu or Link Item Attributes, regardless of which type it is
=> Is  code  in NavigationModelFactory.cs better than NavigationItems.cs?

**Comment by LSNicholls** [comment]
Hi again,

I'll try to answer everything, one item at a time.  Sorry, this will be quite long, because I learned something while trying to do illustrative examples for you and this answer evolved as I worked on it.

1. I understand what you *want* items[i].Title to do.  But for NavigationMenuAttribute items, you also have this:

`  title: strPath,`

... is it possible that these two instructions are incompatible?  What happens if you leave out `  title: strPath,`?  What happens if you put in, instead of   `  title: strPath,`,   `title: "XXXX",`?   In other words, use a literal?

Once again, I have not worked directly with NavigationMenuAttribute and NavigationLinkAttribute, and my impression was that the NavigationItems.cs was supposed to be declarative syntax.  I have never put code there and don't know for sure that you are doing anything wrong.  More on this below.

Also, to prove that the line `items[i].Title = menu[i].Name` is working for the Link items as you expect, you could change it to `items[i].Title ="XXXX"`, again a literal.  Now you should see all the link items with the same caption, which would verify that this line is working and the value isn't coming from anyplace else.

If you don't learn anything from these tests, I did see something interesting with Serenity's menu caching, which *might* be involved, although I don't see how.  **You could try a similar test to what I am showing (changing a second attribute, for me FullPath worked), and see if makes a difference.**  I will explain this fully below.

2. Thank you for showing your Menu table.  My aim here was to understand what `menu[i].Name` might have in it.  I see that CodeName is null for these items, along with Url being null for them -- so I guess neither one is null for the link items.  It probably isn't important/germane to your problem.

3. I am not sure you understood my question about strPath.   But, I am also not sure I understand your code here, ispecially if CodeName is null.  Again, it probably isn't important to your problem.

4.  I think you are saying that of the two classes only one has a title member/property.  I can't speak concretely to that because (again) I don't address these Attribute classes directly and the NavigationModelFactory relies on NavigationHelper, which creates NavigationItems, whether they are menu items or links, they work the same way.

5. & 6.    I am not sure what version of Serenity you are working with, so I can't say that one is "better" than the other, but I have a feeling that the Dynamic Navigation sample might be old.  Typically that means it may be offering a method that is no longer recommended in recent versions.    I have not worked in Serenity before version 6, so if you are working in an older version than that, you may be doing the right thing.  

I will try to describe how it works now.

First I will try to make sure that you understand what I mean about NavigationItems being "declarative".  

In all of the generated code for v6 and above, the *Navigation.cs files generated for various modules only have declarative code in them, which looks like this (where Test is a module and the file for the module is called TestNavigation.cs):

```
using Serenity.Navigation;
using MyPages = MyApp.Test.Pages;

[assembly: NavigationLink(int.MaxValue, "Test/Order Details", typeof(MyPages.OrderDetailsController), icon: null)]
[assembly: NavigationLink(int.MaxValue, "Test/Orders", typeof(MyPages.OrdersController), icon: null)]
[assembly: NavigationLink(int.MaxValue, "Test/Customers", typeof(MyPages.CustomersController), icon: null)]
[assembly: NavigationLink(int.MaxValue, "Test/Products", typeof(MyPages.ProductsController), icon: null)]
```
 
For the Administration module, the AdministrationNavigation.cs file looks like this:

```
using Serenity.Navigation;
using Administration = MyApp.Administration.Pages;
[assembly: NavigationMenu(9000, "Administration", icon: "fa-tools")]
[assembly: NavigationLink(9500, "Administration/Languages", typeof(Administration.LanguageController), icon: "fa-comments")]
[assembly: NavigationLink(9600, "Administration/Translations", typeof(Administration.TranslationController), icon: "fa-comment-o")]
[assembly: NavigationLink(9700, "Administration/Roles", typeof(Administration.RoleController), icon: "fa-lock")]
[assembly: NavigationLink(9800, "Administration/User Management", typeof(Administration.UserController), icon: "fa-users")]

```

Notice that the AdministrationNavigation.cs file has one NavigationMenu() line and several NavigationLink() lines.  **For both types, you can set a title or caption**, although it is done differently for each in this declarative syntax.  For the Menu item, as you see, there is a string ("Administration") without an attribute name, but for the Link items, you can use Title="".  

I am sure that these capabilities correspond, underneath, to the abilities specified by NavigationLinkAttribute versus NavigationMenuAttribute.  

The NavigationItems.cs file is similarly declarative only, and it serves to group things together on the menu, like this:

```
using Serenity.Navigation;

[assembly: NavigationLink(1000, "Dashboard", url: "~/", permission: "", icon: "fa-tachometer")]
[assembly: NavigationSection("Administration/Security",
    Include = new[] { "Administration/Roles", "Administration/User Management" })]

```

The NavigationModelFactory is where all this stuff gets put together, and **this is where the coding is done.**  

Yes, I do think this is better, assuming it is available to you.

It's definitely important to realize that NavigationModelFactory is where the menu *caching* is done. When you figure out your actual method. You will not want to run this code over and over if at all possible.  That's one reason I think that it is better to make the changes in there, or in the Helper class that it invokes.

You will see this in the NavigationModelFactory Create method:

```
  var model = new NavigationModel
  {
      Items = cache.GetLocalStoreOnly("LeftNavigationModel:NavigationItems:" +
          (userAccessor.User?.GetIdentifier() ?? "-1"), 
          TimeSpan.Zero,
          UserPermissionRow.Fields.GenerationKey,
          () => NavigationHelper.GetNavigationItems(permissions, typeSource, serviceProvider, ToAbsolute))
  };

```

You will see that NavigationHelper is providing a list of NavigationItems, and this would include both links and menu items. Again, the NavigationItem class has a Title member, usable for both types.  Your code could leverage getting of the list from NavigationHelper, inside the caching mechanism, like this:

```
   var model = new NavigationModel
   {
       Items = cache.GetLocalStoreOnly("LeftNavigationModel:NavigationItems:" + id,
       //(userAccessor.User?.GetIdentifier() ?? "-1"),
       TimeSpan.Zero,
       UserPermissionRow.Fields.GenerationKey,
       () => {
           List<NavigationItem> myItems = NavigationHelper.GetNavigationItems(permissions, typeSource, serviceProvider, ToAbsolute);
           foreach (var item in myItems)
           {
               ReviseItem(item);
           }
           return myItems;
       })
   };
```

When I did this as a test, I learned a couple of interesting things, which I think might be pertinent to you whether you do this work in the "old" way or the "new" way:

* my proposed RevisedItem method needs to be recursive, so that each menu item reaches down into its child items and makes the right changes for each of them.
* **changes to the Title attribute didn't take unless I made an additional change** (in my case, it worked to change FullPath).  I don't see why this should be and it was definitely the value I gave to Title that was showing, for both menu items and links.  But Title didn't work on its own.  I think this *may* have been something to do with Serenity's menu caching.

Here is my sample ReviseItem code:


```
 private void ReviseItem(NavigationItem item)
 {

     item.FullPath = item.FullPath + " ";  // didn't see results without this!
     item.Title = "AAAA"; // here is where you need to do your lookup from menu

      if (item.Children.Count > 0)
     {
         foreach (var child in item.Children)
         {
            this.ReviseItem(child);
         }
     }
 }
```

Here is the result:

![image](https://github.com/serenity-is/Serenity/assets/88156108/61c26e32-7844-4866-87f4-0bafa9f5fe0a)

So, again, without making other changes, you could try your current method and add a change to FullPath and see if everything magically works then.  But I would still look into making sure you understand the caching of menu items and try to ensure that you aren't doing this work every time every user calls up the menu.

As a final recommendation, when you do custom work like this you usually have to provide a way to invalidate the cache if something changes; for example, if your Menu table is edited.  But this is not difficult, and once you are sure of how you are doing this we can figure that out pretty easily.

**Comment by LSNicholls** [comment, marked as answer]
I need to add further to this... Victor Tomali has explained to me what is going on.  To be sure, I should probably know what version of Serenity you are using, but it is mostly likely to be that _Sidebar.cshtml is looking up the correct Localized value for you, and discarding whatever you put in from your Menu table.

The reason it is working for your link items would be not because link and menu items are truly different, but because your custom items aren't in the localization tables.  But the headers or grouping menu items are not custom, so they are in there and they take precedence.

I think that the text keys for looking up an item in the Translations list (look for the ones that start with Navigation) are based on FullPath, which is why it worked for me when I added a change to FullPath; my items were no longer found in the Translations list.

I suppose this means that, if you're only using the Menu table for these captions, you could add your items to the Translations list and forget about the custom code.  

Alternatively, if your Menu table is handling other things for you, you might want to get rid of the items in the Translations list that represent known headers and are interfering with your custom code.  But I still recommend thinking about caching etc as mentioned earlier.

I hope this helps.  Thanks again to Victor for the "boost".

**Reply by wow-prime** [reply]
I set Navigation text keys for looking up an item in the Translations list, and my problem solved.
I will try your suggestion about caching etc as mentioned earlier in the future.
Thank you very much.
